docs/unified-call-syntax.md

```diff
@@ -0,0 +1,192 @@
+
+```ts
+// This stores a statically known function in `f`.
+let f = Foo::bar;
+
+// This calls Foo::bar(4).
+f 4;
+
+// This also calls Foo::bar(4).
+Foo::bar(4);
+```
+
+## Unqualified Reference
+
+An unqualified function name `identifier` (as distinct from a qualified name such as `ident1::ident2`) can be used. If there is a function in the current scope (either defined in the current class or made available without qualification with a `using` directive), this will resolve as if it were a fully-qualified reference.
+
+If there is no static resolution to a function, this will instead resolve as if it were a dynamic unbound selection on the first argument. It's a compile-time error to try to use an undefined function with no arguments.
+
+```ts
+// The function `foo` must be in scope.
+let f = foo;
+
+// If `foo` is in scope, this calls `foo(x)`. If `foo` is not in scope, this
+// calls `x::foo(x)`.
+foo x;
+```
+
+## Fully-Qualified Infix
+
+A fully-qualified reference used in an infix position receives arguments from both the left-hand side and the right-hand side. If either or both arguments are tuples, the tuples are concatenated into a single tuple.
+
+```ts
+// These are all equivalent.
+Foo::bar(w, x, y, z);
+w Foo::bar (x, y, z);
+(w, x) Foo::bar (y, z);
+(w, x, y) Foo::bar z;
+```
+
+## Unqualified Infix
+
+An unqualified reference used in an infix position receives arguments from both sides in the same way as a fully-qualified reference used in an infix position. The unqualified reference is then either resolved in the current scope or looked up on the first argument.
+
+```ts
+// These are all equivalent. If `foo` is in scope, it is used. Otherwise,
+// `w::foo` must exist and is used.
+foo(w, x, y, z);
+w foo (x, y, z);
+(w, x) foo (y, z);
+(w, x, y) foo z;
+```
+
+## Tuples as Arguments
+
+A tuple as an argument to a function is unpacked into a sequence of arguments. In order to pass a tuple in a specific argument position, nest the tuple.
+
+```ts
+// These are equivalent
+f a b c;
+f(a, b) c;
+f a (b, c);
+a f (b, c)
+(a, b) f c;
+a f b c;
+
+// If the second argument is a tuple, nest it.
+f(a, (b, c));
+```
+
+## Default Arguments
+
+> TODO: default arguments are equivalent to arity-based overloading
+
+```ts
+f(g: I23->I32, a: I32 = 4): I32->I32 
+{
+  { x => g (x + a) }
+}
+
+// Greedy or not?
+f (+ ~ 5) 3
+
+f(a: T, b: U = ...): V { ... }
+let g = f a;
+// g: V or g: U->V ?
+
+class Window
+{
+  class Args
+  {
+    _title: String;
+    _width: USize;
+    _height: USize;
+
+    create(): Args
+    {
+      new ("none", 640, 480)
+    }
+
+    title(self: Args & mut, t: String): Args & mut
+    {
+      self._title = t;
+      self
+    }
+
+    ...
+  }
+
+  create(args: Args): Window { ... }
+}
+
+let w = Window Args.width(1000).title("My Application");
+
+```
+
+## Variable Length Arguments
+
+There is no language support for variable length arguments. Instead, use an iterator.
```

Comment:
Perhaps something like "(Recall that type lists, despite their apparent variable length, elaborate into concrete tuples of particular lengths)"?